### PR TITLE
hide mangling logic from readers of graph_op

### PR DESCRIPTION
**Pull request by marcrasi**
This adds an API that deals with graph_op operand demangling, simplifying code that reads the operands, and making it possible to change the mangling logic without disturbing that code too much.

### CODE REVIEW

File: include/swift/SIL/GraphOperationInfo.h [hunk 1]
```diff
@@ -95,48 +96,66 @@ struct GraphOperationInfo {
...
 
-    OperandMarkerKind Kind;
-
-    /// The mangled name as it appears in the graph_op. Mangling rules described
-    /// in OperandMarkerKind.
-    StringRef MangledName;
-
-    OperandMarker(StringRef MangledName);
+    StructuredOperandKind Kind;
+    StringRef Name;
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/SIL/GraphOperationInfo.h]: Yes. It's empty when it's not there.

File: include/swift/SIL/GraphOperationInfo.h [hunk 2]
```diff
@@ -95,48 +96,66 @@ struct GraphOperationInfo {
...
+    union {
+      /// Operand for SOK_Scalar and SOK_Single.
+      SILValue SingleOperand;
+      /// Operands for SOK_List.
+      ArrayRef<Operand> OperandList;
+    };
+
+    StructuredOperand(StructuredOperandKind Kind, StringRef Name,
+                      SILValue SingleOperand)
+        : Kind(Kind), Name(Name), SingleOperand(SingleOperand) {};
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/SIL/GraphOperationInfo.h]: done

File: include/swift/SIL/GraphOperationInfo.h [hunk 3]
```diff
@@ -95,48 +96,66 @@ struct GraphOperationInfo {
...
+    StructuredOperand(StructuredOperandKind Kind, StringRef Name,
+                      ArrayRef<Operand> OperandList)
+        : Kind(Kind), Name(Name), OperandList(OperandList) {};
 
   public:
-    OperandMarkerKind getKind() const {
+    StructuredOperand(const StructuredOperand &other) = default;
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply include/swift/SIL/GraphOperationInfo.h]: yep, removed.

File: lib/AST/GraphOperationInfo.cpp [hunk 4]
```diff
@@ -76,17 +76,56 @@ void GraphOperationInfo::assertWithDump(bool cond,
...
+    nextMarkerPos = remainingMangled.find(',', 1);
+
+    StringRef thisMarker = remainingMangled.substr(0, nextMarkerPos);
+    StringRef thisMarkerName = thisMarker.drop_front(2);
+    assert(thisMarker.size() >= 2 && "marker too short");
+    switch (thisMarker[1]) {
+    case 's':
+      // Push a SOK_Scalar.
+      assert(thisMarkerName.empty() && "SOK_Scalar should not have name");
+      structuredOperands.push_back({SOK_Scalar, StringRef(),
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/AST/GraphOperationInfo.cpp]: Done, but I had to do something a bit odd to be able to keep the `StructuredOperand` constructor private. Do you think it's overkill? I could maybe just make the `StructuredOperand` constructor public.

[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/AST/GraphOperationInfo.cpp]: This is where I got the idea for the odd thing: https://stackoverflow.com/a/17011117

[the remark this answers is not in the document]
Comment (bgogul) [reply lib/AST/GraphOperationInfo.cpp]: Yes, it is an overkill and unnecessarily confusing for not much benefit. why don't make the constructor public or go back to push_back?

File: lib/AST/GraphOperationInfo.cpp [hunk 5]
```diff
@@ -76,17 +76,56 @@ void GraphOperationInfo::assertWithDump(bool cond,
...
+                                    remainingOperands.take_front(0)});
+      break;
+    case 'e':
+      // Extend the OperandList of the curent SOK_List by 1 to include the next
+      // of the inst's operands.
+      assert(structuredOperands.size() > 0 && "list element not in list");
+      assert(structuredOperands.back().Kind == SOK_List &&
+             "list element not in list");
+      assert(thisMarkerName.empty() && "list element should not have name");
+      structuredOperands.back().OperandList = ArrayRef<Operand>(
```
[the remark this answers is not in the document]
Comment (bgogul) [reply lib/AST/GraphOperationInfo.cpp]: Sounds good.

**Review by bgogul — APPROVED** [review]
LGTM. Just remove the ConstructionAllowed class.

File: lib/AST/GraphOperationInfo.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/AST/GraphOperationInfo.cpp]: made it public

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow linux cpu

**Pull request merged**